### ISSUE TITLE
empty DataSource, and panic while creating DataVolume out of it

### ISSUE THREAD

**Issue by kvaps** [kubevirt/containerized-data-importer#2535]
**What happened**:

I trying to create DataVolume with DataImportCron
```
2023/01/09 19:26:58 http: panic serving 10.111.0.28:20576: runtime error: invalid memory address or nil pointer dereference
goroutine 8058060 [running]:
net/http.(*conn).serve.func1()
	GOROOT/src/net/http/server.go:1825 +0xbf
panic({0x17c5560, 0x2b55130})
	GOROOT/src/runtime/panic.go:844 +0x258
kubevirt.io/containerized-data-importer/pkg/apiserver/webhooks.(*dataVolumeValidatingWebhook).validateDataVolumeSourcePVC(0x0?, 0x0, 0x0?, 0x0?)
	pkg/apiserver/webhooks/datavolume-validate.go:401 +0x5a
kubevirt.io/containerized-data-importer/pkg/apiserver/webhooks.(*dataVolumeValidatingWebhook).validateSourceRef(0xc0004747a0, 0xc000990180?, 0xc0006dc2f8, 0xc000990180, 0x0?)
	pkg/apiserver/webhooks/datavolume-validate.go:397 +0x871
kubevirt.io/containerized-data-importer/pkg/apiserver/webhooks.(*dataVolumeValidatingWebhook).validateDataVolumeSpec(0x1cac4e0?, 0xc000a101a0, 0xc000990180, 0xc0006dc2f8, 0x0?)
	pkg/apiserver/webhooks/datavolume-validate.go:154 +0xb1d
kubevirt.io/containerized-data-importer/pkg/apiserver/webhooks.(*dataVolumeValidatingWebhook).Admit(0xc0004747a0, {{{0xc0006a6f20, 0xf}, {0xc000a9c390, 0x13}}, 0xc000a101a0, 0x0})
	pkg/apiserver/webhooks/datavolume-validate.go:544 +0xc6f
kubevirt.io/containerized-data-importer/pkg/apiserver/webhooks.(*admissionHandler).ServeHTTP(0xc000710700, {0x1cc3dd8, 0xc000a4c1c0}, 0xc0009b2700?)
	pkg/apiserver/webhooks/handler.go:124 +0x44a
net/http.(*ServeMux).ServeHTTP(0x3b?, {0x1cc3dd8, 0xc000a4c1c0}, 0xc0009b2700)
	GOROOT/src/net/http/server.go:2462 +0x149
kubevirt.io/containerized-data-importer/vendor/github.com/emicklei/go-restful.(*Container).ServeHTTP(0xc00058e79a?, {0x1cc3dd8?, 0xc000a4c1c0?}, 0x40ee85?)
	vendor/github.com/emicklei/go-restful/container.go:311 +0x1f5
net/http.serverHandler.ServeHTTP({0x1cb6280?}, {0x1cc3dd8, 0xc000a4c1c0}, 0xc0009b2700)
	GOROOT/src/net/http/server.go:2916 +0x43b
net/http.(*conn).serve(0xc0003e8960, {0x1cc4810, 0xc000166d50})
	GOROOT/src/net/http/server.go:1966 +0x5d7
created by net/http.(*Server).Serve
	GOROOT/src/net/http/server.go:3071 +0x4db
2023/01/09 19:27:07 http: panic serving 10.111.0.28:20636: runtime error: invalid memory address or nil pointer dereference
```

**What you expected to happen**:

Created DataVolume out of DataSource

**How to reproduce it (as minimally and precisely as possible)**:

Try to create:

```yaml
---
apiVersion: cdi.kubevirt.io/v1beta1
kind: DataImportCron
metadata:
  name: ubuntu-2204
  namespace: default
spec:
  managedDataSource: ubuntu-2204
  schedule: "*/2 * * * *"
  template:
    spec:
      source:
        registry:
          pullMethod: node
          url: docker://dev-registry.deckhouse.io/sys/deckhouse-oss:92d31ba325ce661deddad46aae171db1d64536053b97da1652586e76-1666192123869
      storage:
        resources:
          requests:
            storage: 10Gi
---
apiVersion: cdi.kubevirt.io/v1beta1
kind: DataVolume
metadata:
  name: disk-mydata2
  namespace: default
spec:
  pvc:
    accessModes:
    - ReadWriteMany
    resources:
      requests:
        storage: 10Gi
    storageClassName: linstor-thindata-r2
    volumeMode: Block
  sourceRef:
    kind: DataSource
    name: ubuntu-2204
```

**Additional context**:
I noticed that automatically created DataSource has no PVC section:

```yaml
# kubectl get datasource -o yaml ubuntu-2204
apiVersion: cdi.kubevirt.io/v1beta1
kind: DataSource
metadata:
  creationTimestamp: "2023-01-09T19:46:50Z"
  generation: 2
  labels:
    app.kubernetes.io/component: storage
    app.kubernetes.io/managed-by: cdi-controller
    cdi.kubevirt.io/dataImportCron: ubuntu-2204
  name: ubuntu-2204
  namespace: default
  resourceVersion: "111194114"
  uid: 8344327d-74e5-48d8-9588-e1d79b647aca
spec:
  source: {}
status:
  conditions:
  - lastHeartbeatTime: "2023-01-09T19:46:50Z"
    lastTransitionTime: "2023-01-09T19:46:50Z"
    message: No source PVC set
    reason: NoPvc
    status: "False"
    type: Ready
  source: {}
```

**Environment**:
- CDI version: 1.55.1
- Kubernetes version (use `kubectl version`): v1.23.15
- DV specification: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

**Comment by kvaps** [comment]
@awels could you help me investigate this, please?

**Comment by akalenyu** [comment]
Thanks for opening the issue!
Yeah, I see what you mean, the datasource is not getting populated until the import completes.
Need to see how we can be more friendly here, meanwhile, you can workaround this if you place
`k wait dataimportcron ubuntu-2204 --for=condition=UpToDate --timeout=5m`
between the dataimportcron & DV creation

**Comment by awels** [comment]
Thinking about this, we should probably not validate the sourceRef in the validation webhook. Instead in the dv controller we should check if the dataSource is complete before creating the matching PVC. We already set the ready condition on the datasource indicating the source PVC is not set yet. So it should be easy enough in the dv controller to check for that condition to be ready before creating the PVC.

https://github.com/kubevirt/containerized-data-importer/blob/main/pkg/controller/dataimportcron-controller.go#L495-L497 is the code that updates the source. And https://github.com/kubevirt/containerized-data-importer/blob/main/pkg/controller/datavolume/clone-controller.go#L1649-L1668 is what sets the source PVC. We should handle that returning an empty PVC for the source.

**Comment by mhenriks** [comment]
obviously, the api server shouldn't crash, and we should fix that, but we cannot move validation out of the webhook as @awels suggested.  We have to validate that the user creating the DataVolume has permission to the target namespace/pvc

**Comment by mhenriks** [comment]
> @mhenriks is that problem related to the [cross-namespace population](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#cross-namespace-data-sources) issue that we discussed in https://github.com/kubevirt/containerized-data-importer/pull/2482#issuecomment-1382490867? How hard it'd be to keep that user context outside the validation webhooks? Anyway, In the meantime opened a small PR (https://github.com/kubevirt/containerized-data-importer/pull/2548) to reject DV creation and avoid the segfault in this case. Feel free to review :)

Problem is that we only have user info in the webhook.  By the time a controller process a resource, that info is gone.

**Comment by alromeros** [comment]
The fix has been merged, should we close the issue?

**Comment by mhenriks** [comment]
fixed in #2548
